fuzz/ml-kem.c

```diff
@@ -0,0 +1,685 @@
+ * public key, then decapsulates it to retrieve the original key. It makes
+ * use of OpenSSL's EVP_PKEY API for encryption and decryption.
+ *
+ * @param[out] buf   Unused output buffer (reserved for future use).
+ * @param[out] len   Unused length parameter (reserved for future use).
+ * @param[in]  key1  Pointer to an EVP_PKEY structure used for key operations.
+ * @param[in]  in2   Unused input parameter (reserved for future use).
+ * @param[out] out1  Unused output parameter (reserved for future use).
+ * @param[out] out2  Unused output parameter (reserved for future use).
+ */
+static void mlkem_encap_decap(uint8_t **buf, size_t *len, void *key1, void *in2,
+                              void **out1, void **out2)
+{
+    EVP_PKEY *key = (EVP_PKEY *)key1;
+    EVP_PKEY_CTX *ctx;
+    unsigned char genkey[128];
+    size_t genkey_len = 128;
+    unsigned char unwrappedkey[2048];
+    size_t unwrappedkey_len = 2048;
+    unsigned char wrapkey[2048];
+    size_t wrapkey_len = 2048;
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to allocate ctx\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_encapsulate_init(ctx, NULL)) {
+        fprintf(stderr, "Failed to init encap context\n");
+        goto err;
+    }
+
+    if (!RAND_bytes(genkey, genkey_len))
+        goto err;
+
+    if (EVP_PKEY_encapsulate(ctx, wrapkey, &wrapkey_len, genkey, &genkey_len) <= 0) {
+        fprintf(stderr, "Failed to encapsulate key\n");
+        goto err;
+    }
+
+    EVP_PKEY_CTX_free(ctx);
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);
+    if (ctx == NULL) {
+        fprintf(stderr, "Failed to create context\n");
+        goto err;
+    }
+
+    if (!EVP_PKEY_decapsulate_init(ctx, NULL)) {
+        fprintf(stderr, "Failed to init decap\n");
+        goto err;
+    }
+
+    if (EVP_PKEY_decapsulate(ctx, unwrappedkey, &unwrappedkey_len,
+                             wrapkey, wrapkey_len) <= 0) {
+        fprintf(stderr, "Failed to decap key\n");
+        goto err;
+    }
+
+err:
+    EVP_PKEY_CTX_free(ctx);
+    return;
+}
+
+/**
+ * @brief Derives a shared secret using the provided key and peer key.
+ *
+ * This function performs a key derivation operation using the given
+ * private key and peer public key. The resulting shared secret is
+ * allocated dynamically and must be freed by the caller.
+ *
+ * @param[in] key The private key used for derivation.
+ * @param[in] peer The peer's public key.
+ * @param[out] shared Pointer to the derived shared secret (allocated).
+ * @param[out] shared_len Length of the derived shared secret.
+ *
+ * @note The caller is responsible for freeing the memory allocated
+ *       for `shared` using `OPENSSL_free()`.
+ */
+static void do_derive(EVP_PKEY *key, EVP_PKEY *peer, uint8_t **shared, size_t *shared_len)
+{
+    EVP_PKEY_CTX *ctx = NULL;
+
+    *shared = NULL;
+    *shared_len = 0;
+
+    ctx = EVP_PKEY_CTX_new_from_pkey(NULL, key, NULL);
```

Comment:
Does this support KEMs?  I expect it to always fail.

Reply:
not at the moment no, but I had written it before I realized, and though it might still be a good exercise, since, with the fuzzer, success or failure matters less than executing different code paths.